README.md
# league-tooltips

> Express middleware for League of Legends tooltips : champions, items, summoner spells, runes, masteries and champion spells.

![stability-stable](https://img.shields.io/badge/stability-stable-green.svg)

[![npm version][version-badge]][version-url]
[![Known Vulnerabilities][vulnerabilities-badge]][vulnerabilities-url]
[![dependency status][dependency-badge]][dependency-url]
[![devdependency status][devdependency-badge]][devdependency-url]
[![downloads][downloads-badge]][downloads-url]

[![NPM][npm-stats-badge]][npm-stats-url]

![Champion tooltip](previews/champion.png)

[more screenshots here](PREVIEWS.md)

## Demo : [here](https://tooltips.lol-item-sets-generator.org/)

[Used in production by _feeder.lol-item-sets-generator.org_](https://github.com/league-of-legends-devs/feeder.lol-item-sets-generator.org/blob/master/src/routes/tooltips.js).

## Compatibility

### Server

Requires Node 4.0.0 or upper.

### Client

**BEWARE !** The client part is not compatible with modern front end frameworks (with dynamic DOM elements). See the [#1](https://github.com/league-of-legends-devs/league-tooltips/issues/1) issue for more informations.

Working on :

* Internet Explorer 11
* Most modern browsers

## Installation

`npm install --save league-tooltips`

## How to use

### Server :
```javascript
const express = require('express');
const leagueTips = require('league-tooltips');

const app = express();

app.use(leagueTips('RIOT_API_KEY', leagueTips.REGIONS.EUROPE_WEST));

app.listen(3000);
```

### Client :
```html
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>Test</title>
    <script src="/league-tips.min.js"></script>
  </head>
  <body>
    <main>
      <div class="league-tooltip" data-champion="103">
        <!-- This div will show the Ahri tooltip on hover (Ahri ID : 103) -->
      </div>
    </main>
  </body>
</html>
```

You must set the `league-tootip` class and one of these `data-*` attributes :

* `data-champion` : `<champion id>`

 e.g. : `data-champion="103"` (Ahri)

* `data-item` : `<item id>`

 e.g. : `data-item="3027"` (Rod of Ages)

* `data-summonerspell` : `<summonerspell id>`

 e.g. : `data-summonerspell="4"` (Flash)

* `data-rune` : `<rune id>`

 e.g. : `data-rune="5311"` (Greater Seal of Critical Chance)

* `data-mastery` : `<mastery id>`

 e.g. : `data-mastery="6111"` (Fury)

* `data-championspell` : `id.key` (`.` is a separator)

 e.g. : `data-championspell="103.Q"`


The value must be the ID (from the [Riot API](https://developer.riotgames.com/)) of the element you want to show. If an element has more than one tag in this list, the list order will define the priority and the first occurrence will be used.

## API

`leagueTips(apiKey, region, options)` is an [express](http://expressjs.com/) [middleware](http://expressjs.com/en/guide/using-middleware.html).

It will return a `function (req, res, next) {}`.

### `apiKey`

**required**

The Riot API key. You can get one [here](https://developer.riotgames.com/).

### `region`

**required**

The location of the servers to query.
Riot has multiple servers across the world at different locations. They are exported in the `REGIONS` constant.

Allowed values :
* `REGIONS.BRAZIL`: `'br'`
* `REGIONS.EUROPE`: `'eune'`
* `REGIONS.EUROPE_WEST`: `'euw'`
* `REGIONS.KOREA`: `'kr'`
* `REGIONS.JAPAN`: `'jp'`
* `REGIONS.LATIN_AMERICA_NORTH`: `'lan'`
* `REGIONS.LATIN_AMERICA_SOUTH`: `'las'`
* `REGIONS.NORTH_AMERICA`: `'na'`
* `REGIONS.OCEANIA`: `'oce'`
* `REGIONS.RUSSIA`: `'ru'`
* `REGIONS.TURKEY`: `'tr'`

### `options`

**optional**, default values :

```javascript
options = {
  url: '/',
  fileName: 'league-tips.min.js',
  cors: { // the CORS default options are only set if the 'cors' option is defined
    origin: '*',
    methods: 'GET,PUT,POST,DELETE',
    headers: 'Content-Type'
  },
  cache: {
    TTL: 60 * 60 * 12,
    redis: { // the Redis default options are only set if the 'redis' option is defined, otherwise : fallback to in-memory caching
      host: 'localhost',
      port: 6379,
      prefix: 'league-tooltips_'
    }
  },
  prod: false
};
```

#### - `url` (String)

**default value** : `'/'` (e.g. : `/tooltips`)

The route where the static files and the datas will be served.

#### - `fileName` (String)

**default value** : `league-tips.min.js`

The name of the Javasript file that will be served to the client.

*e.g.*

`leagueTips('<API_KEY>', '<REGION>', { url: '/tooltips', fileName: 'league-tips.min.js' })`

will serve the file at `/tootips/league-tips.min.js`.

#### - `cors` (Object)

CORS properties. No CORS if this property is undefined.

* `origin` (String), defaults to `'*'` if `cors` is defined
* `methods` (String), defaults to `'GET,PUT,POST,DELETE'` if `cors` is defined
* `headers` (String), defaults to `'Content-Type'` if `cors` is defined

#### - `cache` (Object)

The datas can be cached in two ways : in-memory caching (by default) or with [Redis](https://redis.io).

* `TTL`: (Number), defaults to `60*60*12` (43200 seconds)
* `redis`: // fallback to in-memory caching if no Redis configuration
  * `host`: (String), defaults to `'localhost'`
  * `port`: (Integer), defaults to `6379`
  * `host`: (String), defaults to `'league-tooltips_'`

#### - `prod` (Boolean)

**default value** : `false`

Defines wether the script runs in a production environment or not. Set to `true` for a production environment.

## Language

You can change the language by setting the `leagueTooltips.locale` value.
The locale is the language of the retrieved datas (e.g., `en_US`, `es_ES`, `fr_FR` ...).

## Debug

This module uses the [debug module](https://www.npmjs.com/package/debug).
In order to show the debug logs, you must enable the debugging logs.

* **Server :** you must set the `league-tooltips:*` value in the `DEBUG` environment variable :
 * Linux : `export DEBUG=league-tooltips:*`
 * Windows : `set DEBUG=league-tooltips:*`
* **Browser :** type the following in the console : `leagueTooltips.debug.enable('*')`


## How does it work ?

The express middleware will retrieve the required datas from the Riot API and store them. It will also serve a javascript file that must be executed by the client.

e.g.

```javascript
app.use(leagueTips('RIOT_API_KEY', 'euw', { url: '/tooltips', fileName: 'league-tips.min.js' }));
```
will serve the javascript file with the `/tooltips/league-tips.min.js` route.

The JS file will listen to the `onmouseover` events of all elements that have the `league-tooltip` class and show a tooltip when it triggers.
Every mouse hover will launch a query to the `/tooltips/` api route (if not stored in the browser cache) and retrieve the datas to show in the tooltip.

## TODO

### Release :
- [ ] IE9 compatibility
- [ ] Champion passives
- [ ] Display 'Energy' instead of 'Mana' for manaless champions
- [ ] Allow to choose between setting the data id or the data key in `data-*`
- [ ] **Add the mouseover listener to the newly created tooltips in the DOM** (using [Mutation Observers](https://www.w3.org/TR/2014/CR-dom-20140508/#mutation-observers)) in order to be compatible with modern front end frameworks - see [#1](https://github.com/league-of-legends-devs/league-tooltips/issues/1)
- [ ] Make the API return the proper HTTP error code (usually always 200)
- [ ] Modular client file

### Ideas :
- [ ] Serve the templates in `league-tooltips.min.js` as lodash template in order to prevent direct templates requests to the server
- [ ] Use CSS modules
- [ ] Champion.GG integration ?
- [ ] Serve the rendered tooltip.html file with express ? => { renderServerSide : true }. Otherwise, use [express-state](https://github.com/yahoo/express-state).
- [ ] Custom tooltip.html in middleware config
- [ ] Start the queries before the mouseover event triggers and cache the results
- [ ] Export the client files in a `league-tooltips-client` module with the Universal Module Definition
- [ ] Cancel the client requests if the user no longer hovers the tooltip element (RxJS style ?)

## License

MIT License

Copyright (c) 2016-2018 **Nicolas COUTIN**

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

Anyway, if you earn money on my open source work, I will fucking end you :)

[version-badge]: https://img.shields.io/npm/v/league-tooltips.svg
[version-url]: https://www.npmjs.com/package/league-tooltips
[vulnerabilities-badge]: https://snyk.io/test/npm/league-tooltips/badge.svg
[vulnerabilities-url]: https://snyk.io/test/npm/league-tooltips
[dependency-badge]: https://david-dm.org/ilshidur/league-tooltips.svg
[dependency-url]: https://david-dm.org/ilshidur/league-tooltips
[devdependency-badge]: https://david-dm.org/ilshidur/league-tooltips/dev-status.svg
[devdependency-url]: https://david-dm.org/ilshidur/league-tooltips#info=devDependencies
[downloads-badge]: https://img.shields.io/npm/dt/league-tooltips.svg
[downloads-url]: https://www.npmjs.com/package/league-tooltips
[npm-stats-badge]: https://nodei.co/npm/league-tooltips.png?downloads=true&downloadRank=true
[npm-stats-url]: https://nodei.co/npm/league-tooltips


CHANGELOG.md
# league-tooltips - change log

# 2.0.5 (Jan. 9th 2018)

* Fixed 'for' loop that broke the tooltips
* Updated dependencies

# 2.0.4 (Oct. 3rd 2017)

## **Fixed :**

* fix multiple dependencies vulnerabilities

# 2.0.3 (July 29th 2017)

## **Fixed :**

* fix the caching with Redis that was always set to 'localhost:6379' with the same key prefix that was caused by the Riot API client (see [this issue](https://github.com/ChauTNguyen/kindred-api/issues/22))
* Update GitHub path to the new one in package.json file

# 2.0.2 (July 21th 2017)

## **Changed :**

* Updated dependencies

# 2.0.1 (July 10th 2017)

## **Fixed :**

* Fix main script path
* Fix `prod` config not working

# 2.0.0 (July 7th 2017)

## **BREAKING CHANGES :**

* [Server] Cache : removed `checkperiod` option.
* [Server] Options : removed `protocol` option.

## **Changes :**

* Now uses the V3 version of the Riot API.

## **Features :**

* [Server] Caching with [Redis](https://redis.io). More details in the ['cache' section in the README](https://github.com/league-of-legends-devs/league-tooltips#--cache-object);

# 1.6.3 (June. 29th 2017)

* Debug log of app mount path

# 1.6.2 (June. 26th 2017)

* Fix README docs

# 1.6.1 (June. 25th 2017)

* Fix package.json file typos

# 1.6.0 (June. 25th 2017)

* Fix #3

# 1.5.2 (May. 26th 2017)

* Update dependencies
* Fix license

# 1.5.1 (May. 23th 2017)

* Update dependencies

# 1.5.0 (Nov. 13th 2016)

## **Fixed :**

* [Client] Round the champions stats
* [Server] Fix requested language strings endpoint always returning the same language regardless the given locale
* [Client] Make the tooltip box no longer overflows after it is rendered with the datas
* [Client] Fix unwanted ending '/' on the base route
* [Client] Remove useless debug log
* Multiple code and linting fixes

## **Added :**

* [Client] Use the retrieved languages strings in order to display the proper language
* [Demo] Add Yarn support for the deploy script
* [Demo] Exclude eslint
* [Demo] Fix public folder not found when the demo is ran outside its directory
* [Server] Fix log being called too early

# 1.4.3 (Nov. 9th 2016)

## **Fixed :**

* [Server] The Router no longer keep the API as an attribute for security purposes.

# 1.4.2 (Nov. 8th 2016)

## **Added :**

* [Client] The tooltip box no longer overflows after it is rendered with the datas.

## **Changed :**

* Yarn support
* Use Airbnb Javascript guidelines
* Little code fixes

# 1.4.1 (Oct. 22th 2016)

## **Fixed :**

* [Server] Fix `base` option interfering with the main router, causing wrong base route

# 1.4.0 (Oct. 22th 2016)

## **Added :**

* [Server] Set the locale in the api requests instead of a static value in the middleware config
* [Client] Request the locales to server
* [Server] `/version` route
* [Server] `base` middleware option

## **Changed :**

* `leagueTooltipsDebug` changed to `leagueTooltips.debug`

# 1.3.0 (Oct. 21th 2016)

## **Added :**

* [Server] Use node-cache for server side caching.

# 1.2.2 (Oct. 17th 2016)

## **Added :**

* [Client] Display and error if the datas cannot be retrieved

## **Fixed :**

* [Client] Adjust the tooltip position regarding the cursor position and the viewport size

# 1.2.1 (Oct. 15th 2016)

## **Added :**

* [Client] Tooltip's `z-index` property set to 100

# 1.2.0 (Oct. 15th 2016)

## **Added :**

* [debug module](https://www.npmjs.com/package/debug) added (see [here](https://github.com/league-of-legends-devs/league-tooltips#debug))
* [cors](https://github.com/league-of-legends-devs/league-tooltips#--cors) option

# 1.1.1 (Oct. 15th 2016)

## **Fixed :**

* [Server] Fix "Resource interpreted as Stylesheet but transferred with MIME type" warning message on stylesheets.

# 1.1.0 (Oct. 11th 2016)

## **Added :**

* [Client] Show the tooltip error if the request is errored.
* [Client & Server] Add champion spell tooltip.
* [Client] Show "Display error" if the template rendering failed.
* [Client] Retry patch fetching if the patch is empty or errored.
* [Server] Serve sitemaps.

## **Changed :**

* [Server] Remove the "domain.tld" part of the `options.url` as the `fetch()` cross-origin policy will deny the request to other domains.
* [Client] Change 'data-spell' attribute to 'data-summonerspell' for better clarity.

## **Fixed :**

* [Server] Fix "undefined e" error when throwing errors.

# 1.0.4 (Oct. 10th 2016)

## **Fixed :**

* Fix undefined "sources" variable.

# 1.0.3 (Oct. 10th 2016)

## **Fixed :**

* Fix "forbidden protocol" message if undefined and fix protocol not being used properly in the API.
* Fix undefined url not being set to the default value.

# 1.0.2 (Oct. 10th 2016)

## **Removed :**

* Unpublish demo directory on npm.

# 1.0.1 (Oct. 10th 2016)

## **Fixed :**

* Fix missing index.js file causing require fail.

# 1.0.0 (Oct. 10th 2016)

First publish


src/server/api.js
import Debug from 'debug';
import _ from 'lodash';
import KindredApi from 'kindred-api';

const debug = Debug('league-tooltips:api');

class Api {
  constructor(apiKey, region, { cache = { TTL: 60 * 60 * 12 }, prod = false }) {
    debug('Initializing API');

    this.api = new KindredApi.Kindred({
      key: apiKey,
      defaultRegion: region,
      debug: !prod,
      showKey: true,
      showHeaders: false,
      limits: prod ? KindredApi.LIMITS.PROD : KindredApi.LIMITS.DEV,
      spread: false,
      retryOptions: {
        auto: true,
        numberOfRetriesBeforeBreak: 3,
      },
      timeout: 5000,
      cache: cache.redis
        ? new KindredApi.RedisCache({
          host: cache.redis.host || 'localhost',
          port: cache.redis.port || 6379,
          keyPrefix: cache.redis.prefix || 'league-tooltips_',
        })
        : new KindredApi.InMemoryCache(),
      cacheTTL: {
        STATIC: cache.TTL,
      },
    });

    debug('Initialized API');
  }

  getSources() {
    return {
      champion: async (id, locale) => {
        const config = {
          id: parseInt(id, 10),
          options: {
            tags: ['tags', 'stats', 'image', 'passive', 'spells', 'info'],
            locale,
          },
        };
        const data = await this.api.Static.champion(config);
        return data;
      },
      item: async (id, locale) => {
        const config = {
          id: parseInt(id, 10),
          options: {
            tags: ['tags', 'image', 'gold'],
            locale,
          },
        };
        const data = await this.api.Static.item(config);
        return data;
      },
      summonerspell: async (id, locale) => {
        const config = {
          id: parseInt(id, 10),
          options: {
            tags: ['image', 'rangeBurn', 'cooldownBurn', 'modes'],
            locale,
          },
        };
        const data = await this.api.Static.spell(config);
        return data;
      },
      rune: async (id, locale) => {
        const config = {
          id: parseInt(id, 10),
          options: {
            tags: ['tags', 'image'],
            locale,
          },
        };
        const data = await this.api.Static.rune(config);
        return data;
      },
      mastery: async (id, locale) => {
        const config = {
          id: parseInt(id, 10),
          options: {
            tags: ['ranks', 'image', 'masteryTree'],
            locale,
          },
        };
        const data = await this.api.Static.mastery(config);
        return data;
      },
      championspell: async (spellId, locale) => {
        const spells = ['Q', 'W', 'E', 'R'];
        const id = spellId.split('.')[0];
        const key = spellId.split('.')[1];
        const spellIndex = _.indexOf(spells, key);
        if (spellIndex === -1) {
          throw new Error('Invalid spell key');
        }
        const config = {
          id: parseInt(id, 10),
          options: {
            tags: ['tags', 'stats', 'image', 'passive', 'spells', 'info'],
            locale,
          },
        };
        const data = await this.api.Static.champion(config);
        return data.spells[spellIndex];
      },
    };
  }

  async getPatchVersion() {
    debug('Api.getPatchVersion() call');
    const patches = await this.api.Static.versions();
    const patchVersion = _.first(patches);
    debug('Patch version', patchVersion);
    return patchVersion;
  }

  async getLocale(locale) {
    debug('Api.getLocales() call');
    const locales = await this.api.Static.getLanguageStrings({ locale });
    const localeData = locales.data;
    return localeData;
  }

  /**
   * Gets the datas to render according to the given data type.
   * Data types only have these values :
   * 'champion', 'item', 'summonerspell', 'rune', 'mastery' and 'championspell'.
   * @param {string} dataType Defines the informations to retrieve andto render.
   * @param {string} id String defining an unique data item.
   * e.g.: "103" for a champion, "103.Q" for a champion spell ...
   * @param {string} locale Defines what language strings will be queried.
   */
  async getData(dataType, id, locale = 'en_US') {
    debug('Api.getData() call', dataType, id, locale);

    const dataSources = this.getSources();
    if (!{}.hasOwnProperty.call(dataSources, dataType)) {
      throw new Error(`unknown data type : ${dataType}`);
    }

    const dataQuery = dataSources[dataType];
    const data = await dataQuery(id, locale);

    return {
      dataType,
      id,
      data,
    };
  }
}

export default Api;
